### ISSUE TITLE
Consider adding in-tree support for async with the runtime crate

### ISSUE THREAD

**Issue by mehcode** [sfackler/rust-native-tls#128]
I need async. tls for a project I'm working on and was investigating the async support here. I found `tokio-tls` but I'd like to try and use rustasync/runtime and std futures (as they are soon to be stable).

The binding to tokio for native-tls ( https://github.com/tokio-rs/tokio/blob/master/tokio-tls/src/lib.rs ) doesn't seem that involved and I was wondering if this could be accepted in-tree under a `runtime` feature ( that uses the [`runtime`](https://github.com/rustasync/runtime) crate to be agnostic of the event loop ).

I'd be happy to contribute this if this is something that'd be accepted.

**Comment by sfackler** [comment]
There's no need to use a special crate to do this - tokio-io doesn't actually depend on any of the tokio runtime bits, it just defines AsyncRead and AsyncWrite traits that work with the standard futures crate.

**Comment by mehcode** [comment]
@sfackler You're totally right. I missed that, this makes this even easier than I was thinking. Would you be opposed to adding this in though (using futures v0.3 instead of tokio-io)?

**Comment by sfackler** [comment]
Once the ecosystem stabilizes around an interface we could add support in potentially, yeah.

**Comment by VictorKoenders** [comment]
It looks like the ecosystem is stabilizing around AsyncRead and AsyncWrite:
https://docs.rs/futures-preview/0.3.0-alpha.17/futures/io/trait.AsyncRead.html
https://docs.rs/futures-preview/0.3.0-alpha.17/futures/io/trait.AsyncWrite.html

One of the issues I'm seeing is that tokio has an explicit dependency on `std::io::Read`, but the `futures::io::AsyncRead` does not. This also means that the `TcpStream` of crates like `runtime` are not guaranteed to implement `std::io::Read`:
https://docs.rs/runtime/0.3.0-alpha.6/runtime/net/struct.TcpStream.html

However `native-tls` depends on `S: io::Read + io::Write,`, so if we were to introduce a feature flag for experimental futures support, we might have to introduce a trait alias:
```rust
#[cfg(not(feature = "futures"))]
type InnerTraitConstraint = io::Read + io::Write;
#[cfg(feature = "futures")]
type InnerTraitConstraint = futures::io::AsyncRead + futures::io::AsyncWrite;
```

However, this runs into issues when multiple crates in your dependencies depend on native-tls, where some depend on non-futures, and others depend on futures. It might be easier to introduce `...Async` aliases for the types, so no conflicts occur.

**Comment by CryZe** [comment]
> tokio has an explicit dependency on std::io::Read
tokio is not using Read at all anymore.

But yes, that means that rust-native-tls now implements no traits for hyper-tls or whatever is on top to work with.

A futures feature that removes the support for io::Read / Write is breaking and thus can't be supported.

**Comment by mehcode** [comment]
For those that want something generic here, https://github.com/async-email/async-native-tls is a wrapper around this crate and adds support for either tokio _or_ async-std depending on the selected feature.

This is probably as good as we can get until `std` gets `AsyncRead` and `AsyncWrite`.
